File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    players_dic = {player:idx for idx,player in enumerate(players)}
    for call in callings:
        i = players_dic[call]
        players[i], players[i - 1] = players[i - 1], players[i]
        
        players_dic[players[i]] = i
        players_dic[players[i - 1]] = i - 1
        
    return players
        

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
         HashMap<String, Integer> playerMap = new HashMap<>();
        
        for (int i = 0; i < players.length; i++) {
            playerMap.put(players[i], i);
        }
         for (String call : callings) {
            int i = playerMap.get(call);
            
            if (i > 0) { 
                String temp = players[i];
                players[i] = players[i - 1];
                players[i - 1] = temp;
                
                playerMap.put(players[i], i);
                playerMap.put(players[i - 1], i - 1);
            }
        }
        
        return players;
    }

}